utils/extraction_utils.py

```diff
@@ -0,0 +1,155 @@
+            dataframe to add number of single cells to
+        well_column_name (str):
+            name of column for wells to use for finding single cell count (defaults to "Metadata_Well")
+
+    Returns:
+        pd.DataFrame:
+            pandas dataframe with new metadata column with single cell count
+    """
+    merged_data = (
+        data_df.groupby([well_column_name])[well_column_name]
+        .count()
+        .reset_index(name="Metadata_number_of_singlecells")
+    )
+
+    data_df = data_df.merge(merged_data, on=well_column_name)
+    # pop out the column from the dataframe
+    singlecell_column = data_df.pop("Metadata_number_of_singlecells")
+    # insert the column as the second index column in the dataframe
+    data_df.insert(2, "Metadata_number_of_singlecells", singlecell_column)
+
+    return data_df
+
+
+def add_sc_count_metadata_file(
+    data_path: pathlib.Path,
+    well_column_name: str = "Metadata_Well",
+    file_type: str = "csv.gz",
+):
+    """
+    This function loads in the saved file from Pycytominer or CytoTable (e.g. normalized, etc.), adds the single cell counts for
+    each well as metadata, and saves the file to the same place (as the same file type)
+
+    Args:
+        data_path (pathlib.Path):
+            path to data file to add single cell count on
+        well_column_name (str):
+            name of column for wells to use for finding single cell count (defaults to "Metadata_Well")
+        file_type (str, optional):
+            the file type of the data (options include parquet, csv, defaults to "csv.gz")
+    """
+    # load in data
+    if file_type == "csv.gz":
+        data_df = pd.read_csv(data_path, compression="gzip")
+    if file_type == "parquet":
+        data_df = pd.read_parquet(data_path)
+    if file_type == "csv":
+        data_df = pd.read_csv(data_path)
+
+    # add single cell count as new metadata column
+    data_df = add_single_cell_count_df(
+        data_df=data_df, well_column_name=well_column_name
+    )
+
+    # save updated df to same path as the same file type
+    if file_type == "parquet":
+        data_df.to_parquet(data_path)
+    if file_type == "csv.gz":
+        data_df.to_csv(data_path, compression="gzip")
+    if file_type == "csv":
+        data_df.to_csv(data_path)
+
+
+def load_sqlite_as_df(
+    sqlite_file: str,
+    image_table_name: str = "Per_Image",
+) -> pd.DataFrame:
+    """
+    load in table with image feature data from sqlite file
+
+    Parameters
+    ----------
+    sqlite_file : str
+        string of path to the sqlite file
+    image_table_name : str
+        string of the name with the image feature data (default = "Per_Image")
+
+    Returns
+    -------
+    pd.DataFrame:
+        dataframe containing image feature data
+    """
+    engine = create_engine(sqlite_file)
+    conn = engine.connect()
+
+    image_query = f"select * from {image_table_name}"
+    image_df = pd.read_sql(sql=image_query, con=conn)
```

Comment:
Consider passing a string to `pd.read_sql(con)` instead of building the connection manually. If you do stick with building the connection manually here, make sure you add code to close the connection.

Reply:
@d33bs What would an example of this look like? I found this as a small chunk of code from Pycytominer so I am not too familiar with what you mean here.

Reply:
Hi @jenna-tomkinson - see this [SO answer](https://stackoverflow.com/a/68202654) for a reference on how to do this using only a string and not managing the connection. If you'd like to keep the connection, consider closing it after the data extraction work is complete ([see here](https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.close)).

Reply:
I just made this change here: [ed60424](https://github.com/WayScience/nf1_cellpainting_data/pull/10/commits/ed60424ac96258133a110407426fca97f1fda578)